lib/statefileio.c

```diff
@@ -110,48 +111,72 @@ sf_FIST_delay_on_write();
 //
 
 
-extern int
-is_drbd_device(
+enum {
+    DEVICE_TYPE_NORMAL = 0,
+    DEVICE_TYPE_DRBD = 1,
+    DEVICE_TYPE_NBD = 2
+};
+
+static int
+get_device_type(
       const char *path)
 {
     static const int drbd_major = 147;
+    static const int nbd_major = 43;
 
     struct stat stats;
     if (stat(path, &stats) == -1)
     {
         log_internal(MTC_LOG_ERR, "SF: failed to execute stat (sys %d).\n", errno);
-        return 0;
+        return DEVICE_TYPE_NORMAL;
     }
-    return S_ISBLK(stats.st_mode) && major(stats.st_rdev) == drbd_major;
+    if (S_ISBLK(stats.st_mode))
+    {
+        const int dev_major = major(stats.st_rdev);
+        if (dev_major == drbd_major)
+            return DEVICE_TYPE_DRBD;
+        if (dev_major == nbd_major)
+            return DEVICE_TYPE_NBD;
+    }
+    return DEVICE_TYPE_NORMAL;
+}
+
+extern int
+is_drbd_device(
+      const char *path)
+{
+    return get_device_type(path) == DEVICE_TYPE_DRBD;
 }
 
 static int
 retry_open(
-  int is_drbd)
+  int device_type)
 {
-  return (errno == EROFS && is_drbd) || errno == EAGAIN;
+    return (errno == EROFS && device_type == DEVICE_TYPE_DRBD) || errno == EAGAIN || errno == EINTR;
 }
 
 extern int
 sf_open(
     char *path)
 {
-    int fd = open(path, (O_RDWR | O_DIRECT));
+    const int device_type = get_device_type(path);
+    const int flags = O_RDWR | O_DIRECT | (device_type == DEVICE_TYPE_NBD ? O_SYNC : 0);
+
+    int fd = open(path, flags);
     if (fd >= 0)
     {
       return fd;
     }
 
-    const int is_drbd = is_drbd_device(path);
-    if (retry_open(is_drbd))
+    if (retry_open(device_type))
```

Comment:
maybe keep `const int is_drdb = device_type == DEVICE_TYPE_DRDB` so you don't need to modify retry_open.